src/main/java/software/bernie/techarium/display/container/PipeContainer.java

```diff
@@ -0,0 +1,121 @@
+        this(id, playerInventory, IWorldPosCallable.NULL, PipePosition.createFromNBT(packetBuffer.readNbt()));
+    }
+
+    public PipeContainer(int id, PlayerInventory playerInventory, IWorldPosCallable worldCallable, PipePosition pos) {
+        super(ContainerRegistry.PIPE_CONTAINER.get(), id);
+        worldPosCallable = worldCallable;
+        tilePos = pos;
+        world = playerInventory.player.level;
+
+        final int SLOT_DIFFERENCE = 18;
+        final int posX = 16;
+        final int posY = 113;
+        final int posHotbarY = 171;
+
+        for(int row = 0; row < 3; row++) {
+            for(int column = 0; column < 9; column++) {
+                addSlot(new Slot(playerInventory, 9 + row*9 + column, posX + column*SLOT_DIFFERENCE, posY + row*SLOT_DIFFERENCE));
+            }
+        }
+        for(int column = 0; column < 9; column++) {
+            addSlot(new Slot(playerInventory, column, posX + column*SLOT_DIFFERENCE, posHotbarY));
+        }
+        startTracking();
+    }
+
+    public Optional<PipeTile> getPipeTile() {
+        TileEntity te = world.getBlockEntity(tilePos.getPos());
+        if (te instanceof PipeTile) {
+            return Optional.of((PipeTile) te);
+        }
+        return Optional.empty();
+    }
+
+    protected void startTracking() {
+        Optional<PipeTile> pipeTileOptional = getPipeTile();
+        pipeTileOptional.ifPresent(pipeTile ->
+            {
+                addDataSlot(new IntReferenceHolder() {
+                    @Override
+                    public int get() {
+                        return pipeTile.getConfig().getInputUsableConfig().get(tilePos.getDirection()).getRedstoneControlType().ordinal();
+                    }
+
+                    @Override
+                    public void set(int value) {
+                        pipeTile.getConfig().getInputUsableConfig().get(tilePos.getDirection()).setRedstoneControlType(RedstoneControlType.values()[value]);
+                    }
+                });
+                addDataSlot(new IntReferenceHolder() {
+                    @Override
+                    public int get() {
+                        return pipeTile.getConfig().getOutputUsableConfig().get(tilePos.getDirection()).getRedstoneControlType().ordinal();
+                    }
+
+                    @Override
+                    public void set(int value) {
+                        pipeTile.getConfig().getOutputUsableConfig().get(tilePos.getDirection()).setRedstoneControlType(RedstoneControlType.values()[value]);
+                    }
+                });
+            }
+        );
+
+    }
+
+    @Override
+    public boolean stillValid(PlayerEntity playerIn) {
+        return stillValid(worldPosCallable, playerIn, BlockRegistry.PIPE.getBlock());
+    }
+
+    @Override
+    public ItemStack quickMoveStack(PlayerEntity playerIn, int index) {
+        return ItemStack.EMPTY;
```

Comment:
wat

Reply:
The vanilla method crashes for some reason, atleast that's how I remember it from previous projects